### PR TITLE
[FI]Fix async mocking in test_agent_loop using AsyncMock

### PR DESCRIPTION
**Pull request by dabasarayash17-pixel**
> **Before opening this PR:**
> - Does it target `dev`? PRs against `main` are auto-closed.
> - Is there a linked issue? PRs without one will be closed.
> - Did you read [CONTRIBUTING.md](../blob/dev/CONTRIBUTING.md)?

## What does this PR do?

This PR fixes a test failure caused by incorrect mocking of asynchronous methods in `test_agent_loop.py`.

Previously, `MagicMock` was used to mock methods that are awaited in the code, which resulted in the following error:

TypeError: object MagicMock can't be used in 'await' expression

This PR replaces `MagicMock` with `AsyncMock` for async methods, ensuring compatibility with `await` expressions and allowing the test suite to run successfully
## Related Issue

Fixes #843

## Changes Made

- Updated test_agent_loop.py to replace MagicMock with AsyncMock
- Fixed test assertion to use resolved_key instead of session_key
- Refactored lock cleanup logic in loop.py by moving it to a finally block
- Removed unnecessary @pytest.mark.asyncio decorator
- Updated test to use shared fixtures instead of patching

1. Clone the repository:
   git clone https://github.com/pocketpaw/pocketpaw.git
   cd pocketpaw

2. Install dependencies:
   uv sync

3. Run tests:
   uv run pytest --ignore=tests/e2e

   Observe that tests in `test_agent_loop.py` fail with:
   TypeError: object MagicMock can't be used in 'await' expression

4. Apply the changes from this PR (replace MagicMock with AsyncMock).

5. Run the tests again:
   py -m pytest

6. Verify that:
   - Async-related errors are resolved
   - Tests run successfully without the previous failure
## Evidence of Testing

```
<img width="960" height="504" alt="image" src="https://github.com/user-attachments/assets/f9bae957-b0bc-40ae-951b-97725ad44bd0" />

```

## Checklist

- [x] PR targets `dev` branch (not `main`)
- [x] Linked to an existing issue
- [x] I have run PocketPaw locally and tested my changes
- [x] Tests pass (`uv run pytest --ignore=tests/e2e`)
- [x] Linting passes (`uv run ruff check .`)
- [x] I have added/updated tests if applicable
- [x] No unrelated changes bundled in this PR
- [x] No secrets or credentials in the diff

### CODE REVIEW

**Comment by DevRohit06** [comment]
### Code review

Found 3 issues. Checked for bugs and CLAUDE.md compliance.

---

**1. New test assertion checks the wrong key — passes trivially and proves nothing** (score: 95)

`test_process_message_cleans_up_lock_on_inner_exception` asserts `msg.session_key not in loop._session_locks`, but `_session_locks` is keyed by `resolved_key`, not `session_key`. Because the test never configures `resolve_session_key` on the mock memory object, `_maybe_await(self.memory.resolve_session_key(session_key))` returns a `MagicMock` instance. The lock is stored under that `MagicMock` as the dict key — `"cli:chat1"` is never inserted. The assertion therefore trivially passes regardless of whether the cleanup code actually runs.

To fix, the test must set `mock_memory.resolve_session_key = AsyncMock(side_effect=lambda k: k)` (matching the pattern used in the existing `make_mock_memory()` helper) and assert against the key that was actually used (`resolved_key` / `msg.session_key` after the passthrough).

https://github.com/pocketpaw/pocketpaw/blob/05b62ab98ebbbd1df449dd43fddfefffb3f31cd1/tests/test_agent_loop.py#L704-L741

---

**2. Stray Git submodule pointer committed with no `.gitmodules` entry** (score: 92)

The diff adds a `pocketpaw` entry of type `commit` (mode `160000`) pointing to `a72bd08e98e935f8a1c387031590ec0dcbf89b34`, but the PR contains no `.gitmodules` file. This is a broken submodule reference — likely the result of accidentally running `git add` inside a nested git repository. It will cause `git submodule update` to fail for anyone who clones the repo with `--recurse-submodules`, and it will confuse CI pipelines that traverse the tree. The entry must be removed (`git rm --cached pocketpaw`) before merge.

https://github.com/pocketpaw/pocketpaw/blob/05b62ab98ebbbd1df449dd43fddfefffb3f31cd1/pocketpaw

---

**3. `_maybe_await` in production code is the wrong fix for a test-double problem** (score: 82)

`_maybe_await` uses `inspect.isawaitable` to paper over the fact that a `MagicMock` (non-async) was used to stand in for an `async def` method. The correct fix per `CLAUDE.md` ("Tests use `pytest-asyncio` with `asyncio_mode = "auto"`; async everywhere") and the existing test helper `make_mock_memory()` is to use `AsyncMock` for async collaborators in tests — which is what `make_mock_memory` already does (`resolve_session_key = AsyncMock(...)`). Introducing runtime `isawaitable` dispatch in `loop.py` for a test-isolation concern adds unnecessary complexity and obscures the real contract of `MemoryStoreProtocol.resolve_session_key` (which is, and must remain, a coroutine). Remove `_maybe_await` and revert the call site to `await self.memory.resolve_session_key(session_key)`.

https://github.com/pocketpaw/pocketpaw/blob/05b62ab98ebbbd1df449dd43fddfefffb3f31cd1/src/pocketpaw/agents/loop.py#L80-L86

---

🤖 Generated with [Claude Code](https://claude.ai/code)
<sub>- If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

**Review by DevRohit06 — CHANGES_REQUESTED** [review]
Wrong key assertion, broken git submodule pointer, _maybe_await is wrong fix for production code. See comment.

**Comment by dabasarayash17-pixel** [comment]
Thanks for the detailed review. I’ve addressed the issues:

- Fixed test assertion by using the correct resolved_key
- Removed _maybe_await and restored direct async handling
- Removed accidental submodule reference and cleaned repository structure

Please review the updated changes.
________________________________

**Review by ragini-pandey — COMMENTED** [review]
## Blocking Issues (must be fixed before merge)

### 1. PR targets `main` — should target `dev`

The base branch is `main`, not `dev`. Per CONTRIBUTING.md and the PR template checklist, all PRs must target `dev`. PRs against `main` are auto-closed. The author needs to retarget the branch.

---

### 2. A git submodule was accidentally committed

```diff
diff --git a/pocketpaw b/pocketpaw
new file mode 160000
index 00000000..a72bd08e
--- /dev/null
+++ b/pocketpaw
@@ -0,0 +1 @@
+Subproject commit a72bd08e98e935f8a1c387031590ec0dcbf89b34
```

A `.gitmodules` entry / submodule pointer to the repo itself was added. This is almost certainly unintentional and will corrupt the repository tree for every consumer. It must be removed before this can be considered.

---

### 3. The PR title and description don't match the actual diff

The title says *"Fix async mocking in test_agent_loop using AsyncMock"* and the description talks exclusively about replacing `MagicMock` with `AsyncMock`. But the diff also includes a non-trivial production change to `src/pocketpaw/agents/loop.py` — moving the eager lock-cleanup from the `try` block into a `finally` block. The description should explain that change separately, with its own motivation, risk, and testing notes.

---

## Correctness Issues

### 4. `lock = None` initialisation then `if lock is not None` — the guard is always false on the exception path that matters

```python
lock = None
try:
    async with self._global_semaphore:
        ...
        lock = self._session_locks[resolved_key]
        ...
        async with lock:
            await self._process_message_inner(...)
finally:
    if lock is not None and not lock.locked():
        self._session_locks.pop(resolved_key, None)
        ...
```

If the exception is raised **inside** `async with lock:` (i.e. by `_process_message_inner`), `lock` is assigned and `lock.locked()` returns `False` once the context manager exits — so cleanup runs correctly here.

However, the `not lock.locked()` check was the **original** guard for the non-exception (success) path, to avoid removing a lock that a second waiting coroutine has just acquired. Moving it to `finally` means it now also fires on the exception path, but only when no other coroutine is waiting. That's correct behaviour, but the `lock = None` initialisation exists solely to handle the case where the semaphore itself raises before `lock` is assigned — which `asyncio.Semaphore` essentially never does. A comment explaining this edge case and why `None` is the right sentinel would make the intent clear.

---

### 5. New test uses `@pytest.mark.asyncio` — not needed in this project

```python
@pytest.mark.asyncio
async def test_process_message_cleans_up_lock_on_inner_exception():
```

Per the project's `pytest-asyncio` config (`asyncio_mode = "auto"` in `pyproject.toml`), the decorator is not needed and is an active rule violation per the copilot instructions. Remove it.

---

### 6. New test doesn't exercise the concurrency case the fix is meant to protect

The scenario the `finally` block guards against is: coroutine A holds the lock and raises, while coroutine B is waiting on `async with lock`. The test only checks single-coroutine cleanup. A truly thorough test would spawn two tasks on the same session key and confirm that B can proceed after A fails. At minimum, note this limitation in the test docstring.

---

### 7. Test instantiates `AgentLoop()` without the fixtures already used by the rest of the file

The rest of `test_agent_loop.py` uses the shared `mock_bus`, `mock_memory`, and `mock_settings` fixtures (defined at the top of the file). The new test re-patches everything inline via `patch(...)` context managers, which is duplicated work and harder to maintain. Prefer the shared fixtures:

```python
async def test_process_message_cleans_up_lock_on_inner_exception(
    agent_loop, mock_memory
):
    ...
```

---

## Nit / Style

### 8. PR description instructions template was not cleaned up

The description still contains the boilerplate `- file_path: description of change` line under "Changes Made". It should be removed.

### 9. "How to Test" section uses `py -m pip` / `py -m pytest` instead of `uv`

The project uses `uv` for all dependency management and test running (`uv run pytest --ignore=tests/e2e`). Using `pip` or bare `pytest` can produce different results and confuses contributors. Update the instructions to match the project tooling.

### 10. Checklist boxes are all unchecked

The author has left all checklist items unchecked (`- [ ]`). Even the items that are clearly satisfied (e.g. "No secrets or credentials in the diff") should be checked before requesting review.

**Review by DevRohit06 — CHANGES_REQUESTED** [review]
Thanks for addressing some of the earlier feedback. A few things still need fixing before this can merge:

1. **Remove `@pytest.mark.asyncio`** — The project uses `asyncio_mode = "auto"` in `pyproject.toml`, so the decorator is unnecessary and inconsistent with the rest of the test file.

2. **Use shared fixtures instead of inline patching** — The rest of `test_agent_loop.py` uses `mock_bus`, `mock_memory`, `mock_settings` fixtures. The new test should follow the same pattern rather than re-patching everything with `patch(...)` context managers.

3. **Confirm the submodule entry is removed** — The earlier review flagged an accidental `pocketpaw` submodule pointer (mode `160000`). Please verify it's no longer in the branch.

4. **Update PR description to cover the `loop.py` change** — The `try/finally` refactor in `_process_message` is a real production behavior change (lock cleanup now runs on exception paths too). The description only mentions AsyncMock test fixes. Please document the motivation and risk of this change separately.

5. **"How to Test" should use `uv`** — Replace `py -m pip` / `py -m pytest` with `uv run pytest` to match project tooling.

**Comment by dabasarayash17-pixel** [comment]
Thanks for the detailed and thorough review.

I’ve addressed all the blocking and correctness issues raised:

* Retargeted the PR to the `dev` branch
* Removed the accidental submodule reference and cleaned the repository structure
* Updated the PR title and description to reflect both the async mocking fix and the lock cleanup change in `loop.py`, including motivation and impact
* Added clarification comments around the `lock = None` initialization to explain the edge case handling
* Removed the unnecessary `@pytest.mark.asyncio` decorator
* Updated the test to use shared fixtures (`agent_loop`, `mock_memory`) instead of inline patching
* Added a note in the test regarding the current limitation around concurrency coverage
* Cleaned up the PR description by removing template placeholders
* Updated the "How to Test" section to use `uv` as per project standards
* Completed the checklist items in the PR template

Please review the updated changes. I appreciate the detailed feedback — it helped improve both the correctness and clarity of the implementation.

**Comment by dabasarayash17-pixel** [comment]
Thanks for the detailed and thorough review, sir I will sent these message 3 days ago on github PR comment but after not getting reply I will sent this mail to you I will completed these issues  please check this and get a reply
I’ve addressed all the blocking and correctness issues raised:

  *   Retargeted the PR to the dev branch
  *   Removed the accidental submodule reference and cleaned the repository structure
  *   Updated the PR title and description to reflect both the async mocking fix and the lock cleanup change in loop.py, including motivation and impact
  *   Added clarification comments around the lock = None initialization to explain the edge case handling
  *   Removed the unnecessary @pytest.mark.asyncio decorator
  *   Updated the test to use shared fixtures (agent_loop, mock_memory) instead of inline patching
  *   Added a note in the test regarding the current limitation around concurrency coverage
  *   Cleaned up the PR description by removing template placeholders
  *   Updated the "How to Test" section to use uv as per project standards
  *   Completed the checklist items in the PR template

Please review the updated changes. I appreciate the detailed feedback — it helped improve both the correctness and clarity of the implementation.

________________________________

**Comment by DevRohit06** [comment]
## Re-review — PR is now non-functional: production fix was removed, test still expects it

Thanks for addressing most of the prior feedback. Unfortunately, in the process of cleaning things up, the production fix itself was dropped while the test that depends on it was kept. The test cannot pass against the current diff.

### Critical — the production change is gone from `loop.py`

The entire `loop.py` diff in this PR now reads:

```diff
-        self.context_builder.bootstrap = DefaultBootstrapProvider()
+        self.context_builder.bootstrap = DefaultBootstrapProvider()
\ No newline at end of file
```

That's not a real code change — it just removes the trailing newline. The `try/finally` refactor you still describe in the PR body (*"Refactored lock cleanup logic in loop.py by moving it to a finally block"*) is no longer present in the diff.

### The new test expects behavior that doesn't exist in `dev`

`test_process_message_cleans_up_lock_on_inner_exception` asserts:

```python
assert resolved_key not in agent_loop._session_locks
assert resolved_key not in agent_loop._session_lock_last_used
```

…after `_process_message_inner` raises. But current `_process_message` in `dev` (`src/pocketpaw/agents/loop.py:335-370`) does **not** clean up locks in a `finally` block. Tracing the exception path:

1. `lock = self._session_locks[resolved_key]` — lock is stored in the dict.
2. `async with lock:` — acquired.
3. `await self._process_message_inner(...)` — **raises**.
4. `async with lock:` exits (lock released).
5. Exception propagates **past** the `if not lock.locked(): self._session_locks.pop(resolved_key, None)` at `loop.py:364` — that cleanup is never reached.
6. Exception propagates past `async with self._global_semaphore:`.
7. `except asyncio.CancelledError:` does **not** catch a generic `Exception`.
8. Exception propagates to the caller — with the lock still in `_session_locks`.

The inline comment on `loop.py:361` even acknowledges this: *"The GC task is a safety net for the cases where this eager path is skipped (e.g. after an exception propagates past this block)."*

So when the test runs: the lock is inserted, the exception fires, cleanup never runs, `resolved_key in agent_loop._session_locks` is **`True`**, and the assertion fails.

In short: **this test will fail against the current PR**. Either it wasn't actually run, or it was run on a working copy that still had the `loop.py` refactor you later removed.

### Please verify

Before re-requesting review, please run:

```bash
uv run pytest tests/test_agent_loop.py::test_process_message_cleans_up_lock_on_inner_exception -v
```

…against a clean checkout of this branch, and paste the **text** output. The current "Evidence of Testing" in the PR body is a broken image tag with no alt text and doesn't reflect the current commit.

### Trailing-newline regression

The new state of `loop.py` ends with `No newline at end of file`. Pre-commit / ruff `W292` will flag this. Please put the newline back — no part of this PR should be touching that line in the first place.

### Credit where it's due — real progress since my last review

To be fair, most of the earlier feedback *was* addressed:

- ✅ Accidental `pocketpaw` submodule pointer (mode 160000) is no longer in the diff
- ✅ Test assertion uses `resolved_key`, and `mock_memory.resolve_session_key = AsyncMock(return_value=resolved_key)` is set correctly — the "wrong key" issue is fixed
- ✅ `_maybe_await` is gone from `loop.py`
- ✅ `@pytest.mark.asyncio` decorator removed
- ✅ New `agent_loop` fixture added and used — matches the file's existing pattern

That's real progress. The problem is that in backing out `_maybe_await` and the `_process_message` refactor together, the *correct* version of the production fix got dropped too.

### Path forward

**1. Put the production fix back in `loop.py`**, using a `try/finally` so lock cleanup runs on every exit path:

```python
async def _process_message(self, message: InboundMessage) -> None:
    session_key = message.session_key
    resolved_key = await self.memory.resolve_session_key(session_key)
    lock: asyncio.Lock | None = None
    try:
        async with self._global_semaphore:
            if resolved_key not in self._session_locks:
                self._session_locks[resolved_key] = asyncio.Lock()
            lock = self._session_locks[resolved_key]
            self._session_lock_last_used[resolved_key] = time.monotonic()
            async with lock:
                await self._process_message_inner(message, resolved_key)
    except asyncio.CancelledError:
        logger.info("Processing cancelled for session %s", session_key)
        raise
    finally:
        # Run on both success and exception paths. Only pop if no other
        # coroutine is currently waiting on this lock, to avoid a race
        # where a waiter re-inserts it immediately after the pop.
        if lock is not None and not lock.locked():
            self._session_locks.pop(resolved_key, None)
            self._session_lock_last_used.pop(resolved_key, None)
```

The `lock: asyncio.Lock | None = None` sentinel handles the (effectively impossible but technically reachable) case where `async with self._global_semaphore:` raises before `lock` is assigned. Please add a short comment explaining that — @ragini-pandey raised the same concern in review point #4.

**2. Restore the trailing newline on `loop.py`.**

**3. Run the test suite locally and paste real text output** (not an image) as the Evidence of Testing:

```bash
uv run pytest tests/test_agent_loop.py -v
uv run ruff check .
uv run mypy src/pocketpaw/agents/loop.py
```

**4. Consider @ragini-pandey's concurrency-coverage concern (review point #6).** A truly thorough test would spawn two tasks on the same session key, have the first one raise, and assert the second one can still proceed. Not a blocker, but worth adding a docstring note if you don't want to extend the test.

**Review by DevRohit06 — CHANGES_REQUESTED** [review]
Re-review: the production fix (try/finally in _process_message) was removed from loop.py but the test that depends on it was kept — the new test `test_process_message_cleans_up_lock_on_inner_exception` will FAIL against the current diff because dev's `_process_message` does not run lock cleanup on the exception path (src/pocketpaw/agents/loop.py:364 is skipped when the inner method raises). Also: loop.py is missing its trailing newline now (W292). Credit for fixing the submodule, wrong-key assertion, `_maybe_await`, `@pytest.mark.asyncio`, and fixture pattern — that's all solid. See comment for the full path forward.

**Comment by dabasarayash17-pixel** [comment]
Hi Rohit Kushwaha,
I’ve pushed the latest updates to the PR addressing the previous review comments.
Key changes:

  *   Restored the try/finally lock cleanup logic in loop.py to ensure proper cleanup on all execution paths
  *   Fixed mypy issues related to AsyncIterator.aclose and set.add
  *   Restored the trailing newline to satisfy lint checks

I verified the changes on a clean setup using uv run:

  *   All tests are passing
  *   No linting issues (ruff)
  *   No type-checking errors (mypy)

I’ve also updated the PR description with the full test output under “Evidence of Testing”.
Could you please review it when you have time? I’d really appreciate your feedback.
Thanks again for your guidance!
Best regards,
Yash

________________________________

**Comment by dabasarayash17-pixel** [comment]
Follow-up on Job Application Status

Dear Sir/Madam,
I hope you are doing well.
I am writing to follow up on my application for the position at your organization. I had previously sent an email around five days ago regarding the status of my application, but I have not yet received any update.
I understand that the selection process may take time; however, I would greatly appreciate it if you could provide me with an update regarding my application status. I am very interested in this opportunity and have been eagerly waiting for your response.
Please let me know if any additional information or changes are required from my side. I would be happy to provide them promptly.
Thank you for your time and consideration. I look forward to your response.
Kind regards,
Yash Dabasara
________________________________

**Comment by prakashUXtech** [comment]
Conflicting with dev. Also worth checking against #822, which is in the same AsyncMock territory — if there's overlap, one of the two should be closed. Rebase, flag the overlap if any, and I'll review.

**Comment by DevRohit06** [comment]
Closing as part of a stale-PR cleanup (PR has been open >2 weeks with unaddressed review feedback). The change is small enough that if it is still applicable against current `dev`, we can pick it up directly. Thanks for the contribution — please feel free to rebase and reopen if you believe it is still needed.
